lib/runner.js

```diff
@@ -408,16 +431,30 @@ class Runner extends Emittery {
 				continue;
 			}
 
+			if (this.runOnlySelectedLines && !task.metadata.selected) {
+				continue;
+			}
+
 			this.emit('stateChange', {
 				type: 'selected-test',
 				title: task.title,
 				knownFailing: false,
 				skip: false,
 				todo: true
 			});
+
+			todoTests.push(task);
+		}
+
+		const hasRunnableTests = concurrentTests.length > 0 || serialTests.length > 0;
+
+		const hasSelectedTests = hasRunnableTests || todoTests.length > 0;
+		if (this.runOnlySelectedLines && !hasSelectedTests) {
+			this.emit('error', 'No tests selected by line numbers.');
```

Comment:
We should handle this the same way we handle `--match`, in the reporters. But only if line numbers were specified for this test file:

https://github.com/avajs/ava/blob/cb5f9f70879ee549837c31e577e01f87970164ed/lib/reporters/mini.js#L388

https://github.com/avajs/ava/blob/cb5f9f70879ee549837c31e577e01f87970164ed/lib/reporters/verbose.js#L330

Reply:
Sounds good!

Reply:
Not quite sure how to do this efficiently, though. In the `--match` case we pass `matching` as a property of the plan when the run is started. The `run` event is emitted before each file is being processed in `api.js`. 

We would need to compute a map from file to `hasLineNumbers/runOnlySelectedLines` and pass down? How do we know which test file is running in the reports, though? 

I noticed that some hooks receive event data with the currently running test, but `reporter.endRun()` doesn't seem to receive any such info and as far as I can tell no such state is saved in the reporters.

Reply:
You're right `matching` is for the global `--match` option.

From the API, the run status for each file is set up here:

https://github.com/avajs/ava/blob/18aeac64db2711d37853175b5e8dc1fd72c66d2a/lib/api.js#L226

You could pass whether we're hoping to select lines. Store it here:

https://github.com/avajs/ava/blob/18aeac64db2711d37853175b5e8dc1fd72c66d2a/lib/run-status.js#L34

This then becomes available in `fileStats` here:

https://github.com/avajs/ava/blob/18aeac64db2711d37853175b5e8dc1fd72c66d2a/lib/reporters/verbose.js#L196

Look for `selectedTests === 0`.

Reply:
Just to add, I pointed you in the wrong direction with `endRun()`. That's for the entire test run, because `--match` applies to the whole thing. Here we want the `worker-finished` event.

Reply:
Alright, wow. Not sure I would've been able to figure that out on my own... Thanks!

Reply:
Something like that?

I assumed you wanted the `mini` reporter to be updated as well accordingly? Question: Do you want a `filesWithoutSelectedTests` as well?

Reply:
Also, it felt a little weird to add the boolean to the stats object, but you lead, I follow.

Reply:
> Alright, wow. Not sure I would've been able to figure that out on my own... Thanks!

Yes it's not great… but it works!

> I assumed you wanted the `mini` reporter to be updated accordingly as well?

Yes. Sorry about the duplicated code.

> Question: Do you want a filesWithoutSelectedTests as well?

Not sure what you mean by this?

Reply:
> Not sure what you mean by this?

I'm referring to this guy: https://github.com/avajs/ava/blob/18aeac64db2711d37853175b5e8dc1fd72c66d2a/lib/reporters/mini.js#L209

Reply:
Ah. Will have to dig into this.